[poppler] Build and package private/unstable headers

nephros opened the pull request:
Poppler can optionally build and install so-called private or unstable header files.

This switches this on and packages the results.

nephros commented:
At least one (upcoming) package in Chum needs these as abuild dependency.

Due to chum policy we can't build the package there so we'd like upstream to provide them.

nephros commented:
Briefly discussed in the Meeting here:

https://irclogs.sailfishos.org/meetings/sailfishos-meeting/2024/sailfishos-meeting.2024-02-15-08.00.log.html

pvuorela commented on rpm/poppler.spec:
```diff
@@ -142,6 +153,12 @@ rm -f %{buildroot}%{_libdir}/*.la
 %{_libdir}/libpoppler-cpp.so
 %{_includedir}/poppler
 
+%files private-devel
```
The above devel package includes already %{_includedir}/poppler so it has all the content that's listed below. Should %exclude there what's not supposed to go to the package.

I'm noticing there's already a similar problem with the qt5-devel subpackage.

Thaodan replied:
Do we have to have a separate package for these? Neither Fedora nor openSUSE have separate packages for these. 

These seem to be entirely opt-in.

nephros replied:
> The above devel package includes already %{_includedir}/poppler so it has all the content that's listed below. Should %exclude there what's not supposed to go to the package.
> 
> I'm noticing there's already a similar problem with the qt5-devel subpackage.

Right, sorry about that. I shall rework the packaging.

But in general, building and packaging these private headers would be okay by you?

nephros replied:
> Do we have to have a separate package for these? Neither Fedora nor openSUSE have separate packages for these.
> 
> These seem to be entirely opt-in.

It has the advantage that one (as a dev building on poppler) has to explicitly require the -private package in order to use them.

pvuorela replied:
This starts to be getting a bit complicated, and I wouldn't be that excited on going through the file list whenever updating poppler. Maybe we should just include these in the main -devel package. There doesn't seem to be other side-effects than installation of the files, and app developer should have some idea from other places on what APIs should be used.

Out of curiousity, what's the thing or feature that needs the unstable APIs? The stable ones should be having quite a lot of features supported already.

nephros replied:
> This starts to be getting a bit complicated, and I wouldn't be that excited on going through the file list whenever updating poppler. Maybe we should just include these in the main -devel package. There doesn't seem to be other side-effects than installation of the files, and app developer should have some idea from other places on what APIs should be used.

Right, okay that lines up with what @Thaodan said earlier.

> Out of curiousity, what's the thing or feature that needs the unstable APIs? The stable ones should be having quite a lot of features supported already.

KDE library 'kitinerary' (not to be confused with KDE Itininerary, the application).
https://invent.kde.org/pim/kitinerary/-/blob/master/CMakeLists.txt?ref_type=heads#L73

Haven't looked into the code but @attah said on IRC it's probably inspection of PDF internals.

attah replied:
In all fairness the list of official .h files is very short.
I really liked the package split (hoping the official part can be OK:ed in Harbour someday, and the split being an extra help to not accidentally use private apis).
But after all it is your packaging.

nephros replied:
The files list was generated by condensing the relevant section in CMakeLists.txt.

So a variant is to use a precheckin script to generate that list, and use it in the RPM (or do the whole thing in the %install section).

Like: 

 - https://github.com/nephros/sailfish-poppler/blob/308fd5e284bad0db80910a14ca5af484f306512f/rpm/poppler.spec#L124-L131
 - https://github.com/nephros/sailfish-poppler/blob/308fd5e284bad0db80910a14ca5af484f306512f/rpm/precheckin_generate_headerlist.sh

Not exactly reducing complexity though.

nephros committed:
[poppler] Package for private/unstable headers

 - Use https schema
 - Fix header packaging
 - Package private headers
 - Clean up %defattr lines
 - use %cmake_foo macros

nephros commented:
New version:

  - Relocate packaging of headers for all -devel packages
  - Use external list for -private-headers (condensed from [this section](https://github.com/sailfishos-mirror/poppler/blob/b58a49e4e576492219b87ed09ed8e02d565d202c/CMakeLists.txt#L669) of the CMakeList

Build artefacts [here](https://build.sailfishos.org/package/show/home:nephros:devel:qt6/poppler-unstable-header).

nephros commented:
^^ Put all the new headers into -devel package

pvuorela commented on rpm/poppler.spec:
```diff
@@ -140,7 +141,13 @@ rm -f %{buildroot}%{_libdir}/*.la
+%{_includedir}/poppler/*.h
+%{_includedir}/poppler/goo/
+%{_includedir}/poppler/fofi/
+%{_includedir}/poppler/splash/
```
Instead of avoiding the earlier %{_includedir}/poppler and having a fine-grained list here,the same thing could be achieved by having here

%exclude %{_includedir}/qt5
%exclude %{_includedir}/glib

pvuorela replied:
This still.

pvuorela commented on rpm/poppler.spec:
```diff
@@ -126,7 +127,7 @@ rm -f %{buildroot}%{_libdir}/*.la
+%{_includedir}/poppler/qt5/
 
 %files
 %defattr(-,root,root,-)
```
If you feel like it, these %defattrs can be removed from the file entries. That's the default already, and we've been cleaning them up.

Thaodan replied:
Also make sure the -devel packages own `%{_includedir}/poppler`.

pvuorela commented:
Please rebase and squash the commits.

Thaodan commented:
Also it would make sense to change the package to use `%cmake_build` and `%cmake_install`.

nephros commented:
While we're cleaning up the little things: remove the man pages as well?

pvuorela commented on rpm/poppler.spec:
```diff
@@ -119,42 +120,44 @@ rm -f %{buildroot}%{_libdir}/*.la
 %{_libdir}/libpoppler-qt5.so
 %{_libdir}/pkgconfig/poppler-qt5.pc
-%{_includedir}/poppler/qt5
+%dir %{_includedir}/poppler
```
Don't think there's any reason for this? It's already owned by -devel and qt5-devel depends on it.

Thaodan replied:
It still has to be owned by the package. It's one of these things that will be errors with RPM 4.20.

See here for example the openSUSE package below if I remove the exact same line in their poppler package:
https://build.opensuse.org/package/live_build_log/home:Thaodan:branches:X11:common:Factory/poppler:qt6/openSUSE_Tumbleweed/x86_64

pvuorela replied:
It is already owned by a package. This is a duplicate entry.

nephros commented:
Edited to *not* own the includedir in the foo-devel packages, and squashed.

As briefly discussed in https://irclogs.sailfishos.org/meetings/sailfishos-meeting/2024/sailfishos-meeting.2024-10-31-08.01.log.html

mlehtima commented on rpm/poppler.spec:
```diff
@@ -119,42 +120,39 @@ rm -f %{buildroot}%{_libdir}/*.la
 %{_libdir}/libpoppler-cpp.so
-%{_includedir}/poppler
+%{_includedir}/poppler/
+%{_includedir}/poppler/cpp/
```
This line is useless, included already by the above line

pvuorela reviewed (approved):
Looking nice and clean now. Thanks!
